sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1481,19 +1489,12 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
       builder.putString("comment", string(STRING))
     }
     // Add Hive type string to metadata.
-    dataType match {
-      case p: PrimitiveDataTypeContext =>
-        p.identifier.getText.toLowerCase match {
-          case "varchar" | "char" =>
-            builder.putString(HIVE_TYPE_STRING, dataType.getText.toLowerCase)
-          case _ =>
-        }
-      case _ =>
-    }
+    val rawDataType = typedVisit[DataType](ctx.dataType)
+    builder.putString(HIVE_TYPE_STRING, rawDataType.catalogString)
```

Comment:
the tests failed because we always put the hive type string. We should only do it when there is char/varchar type